xdsl/backend/riscv/lowering/convert_memref_to_riscv.py

```diff
@@ -0,0 +1,135 @@
+from collections.abc import Sequence
+from typing import Any, cast
+
+from xdsl.backend.riscv.lowering.utils import (
+    cast_operands_to_regs,
+    register_type_for_type,
+)
+from xdsl.dialects import memref, riscv
+from xdsl.dialects.builtin import (
+    ModuleOp,
+    UnrealizedConversionCastOp,
+)
+from xdsl.ir import MLContext, SSAValue
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    GreedyRewritePatternApplier,
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+from xdsl.utils.exceptions import DiagnosticException
+
+
+class ConvertMemrefAllocOp(RewritePattern):
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: memref.Alloc, rewriter: PatternRewriter) -> None:
+        raise DiagnosticException("Lowering memref.alloc not implemented yet")
+
+
+class ConvertMemrefDeallocOp(RewritePattern):
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: memref.Dealloc, rewriter: PatternRewriter) -> None:
+        raise DiagnosticException("Lowering memref.dealloc not implemented yet")
+
+
+def insert_shape_ops(
+    mem: SSAValue,
+    indices: Sequence[SSAValue],
+    shape: Sequence[int],
+    rewriter: PatternRewriter,
+) -> SSAValue:
+    """
+    Returns ssa value representing pointer into the memref at given indices.
+    """
+    assert len(shape) == len(indices)
+
+    match indices:
+        case [idx1]:
+            rewriter.insert_op_before_matched_op(
+                [
+                    ptr := riscv.AddOp(mem, idx1),
+                ]
+            )
+        case [idx1, idx2]:
+            rewriter.insert_op_before_matched_op(
+                [
+                    cols := riscv.LiOp(shape[1]),
+                    row_offset := riscv.MulOp(cols, idx1),
```

Comment:
note: mul is not an op from the base ISA. We might want to make this opt-in at some point.
In snitch cluster, one M extension ALU is shared by all cores in a compute cluster.

Reply:
Interesting, I'm not sure exactly what you mean, but I'm guessing I'll find out soon enough once we try to run this on Snitch and it fails :)

Reply:
`RV32I`, the most basic ISA to run a RISCV program does not have a multiply instruction. Rather, the compiler unrolls it into a loop with multiple add instructions or converts it into a shift instruction of it's a multiple of 2.

If you have `RV32IM` you have a dedicated instruction and a dedicated multiplier that performs the addition of the partialsums of the multiplication in hardware.

Since snitch is mainly a floating point machine, the integer multiplication unit (the on that provides M) is shared across all compute cores. This should work, but you can not parallelize it across multiple cores, or you will stall certain cores, until the shared multiplier is available. If you just consider single-core it should not be a problem at all. Just note that this will not work on very simple RISCV-cores (ones that don't support M extension).

Reply:
In principle people could disable the snitch's multiplier.
just a nit.

The ISA of the compute cores is set here: https://github.com/pulp-platform/snitch_cluster/blob/772b86ae84ec0d5a6f1e755cb524ba0aae2cefc3/target/snitch_cluster/cfg/default.hjson#L86